Require MallocPtr parameter type to be trivially destructible, correctly destroy FuncRefTable

ast-hugger opened the pull request:
#### a892900ddc2177f2283f274756a5816d8a5208f1
<pre>
Require MallocPtr parameter type to be trivially destructible, correctly destroy FuncRefTable
<a href="https://bugs.webkit.org/show_bug.cgi?id=294022">https://bugs.webkit.org/show_bug.cgi?id=294022</a>
<a href="https://rdar.apple.com/152575154">rdar://152575154</a>

Reviewed by Keith Miller.

This patch changes `MallocPtr` to require the parameter type `T` to be trivially destructible.

The usage of `MallocPtr` in `WasmTable` which was responsible for a callee leak is now
incompatible with that requirement. `WasmTable` is changed to use `std::unique_ptr`
with a custom deleter that correctly destroys the table contents.

Additionally, `VM` is changed to use a `FixedVector` instead of a `MallocPtr` to allocate
the exception fuzzing buffer.

* Source/JavaScriptCore/bytecode/ExpressionInfo.h:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::exceptionFuzzingBuffer):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::reallocate):
(JSC::Wasm::Table::grow):
(JSC::Wasm::ExternOrAnyRefTable::ExternOrAnyRefTable):
(JSC::Wasm::FuncRefTable::FuncRefTable):
(JSC::Wasm::FuncRefTable::~FuncRefTable): Deleted.
* Source/JavaScriptCore/wasm/WasmTable.h:
(JSC::Wasm::Table::StorageDeleter::operator()):
* Source/WTF/wtf/MallocPtr.h:

Canonical link: <a href="https://commits.webkit.org/297550@main">https://commits.webkit.org/297550@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cd5055032438f3fa6750f9068995bd5170df720

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21346 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117260 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61497 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31576 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39477 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84547 "Found 3 new test failures: imported/w3c/web-platform-tests/wasm/core/linking.wast.js.html imported/w3c/web-platform-tests/wasm/core/table_copy.wast.js.html imported/w3c/web-platform-tests/wasm/core/table_init.wast.js.html (failure)") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100140 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64993 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24557 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18281 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61080 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103720 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94593 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120358 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109782 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38278 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28444 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93479 "Found 3 new test failures: imported/w3c/web-platform-tests/wasm/core/linking.wast.js.html imported/w3c/web-platform-tests/wasm/core/table_copy.wast.js.html imported/w3c/web-platform-tests/wasm/core/table_init.wast.js.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38654 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96417 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93304 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/23960 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38389 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16158 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34263 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18075 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38167 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43644 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134058 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37832 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36151 "Found 207 new JSC stress test failures: microbenchmarks/wasm-cc-int-to-int.js.default-wasm, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.wasm-bbq, wasm.yaml/wasm/extended-const-spec-tests/elem.wast.js.wasm-bbq-no-consts, wasm.yaml/wasm/function-tests/context-switch.js.default-wasm, wasm.yaml/wasm/function-tests/context-switch.js.wasm-bbq, wasm.yaml/wasm/function-tests/context-switch.js.wasm-bbq-no-consts, wasm.yaml/wasm/function-tests/context-switch.js.wasm-collect-continuously, wasm.yaml/wasm/function-tests/context-switch.js.wasm-eager, wasm.yaml/wasm/function-tests/context-switch.js.wasm-eager-jettison, wasm.yaml/wasm/function-tests/context-switch.js.wasm-no-cjit ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41165 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39534 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6678b995c7d97a11ea64a5fb138240d3c5076524. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46668

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bc85f992fb8f3f23e62dfebbdeb52df9083bb3e7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46668

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTable.h:
```diff
@@ -52,8 +51,18 @@ class Table : public ThreadSafeRefCounted<Table> {
+        void operator()(T* ptr);
+    };
+
     JS_EXPORT_PRIVATE ~Table() = default;
```
Can we move this line next to the `operator delete` overload. It would make it a bit clearer what's going on.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTable.cpp:
```diff
@@ -263,28 +293,23 @@ FuncRefTable::FuncRefTable(uint32_t initial, std::optional<uint32_t> maximum, Ty
     // FIXME: It might be worth trying to pre-allocate maximum here. The spec recommends doing so.
     // But for now, we're not doing that.
     // FIXME this over-allocates and could be smarter about not committing all of that memory https://bugs.webkit.org/show_bug.cgi?id=181425
+    uint32_t actualLength = allocatedLength(m_length);
```
The fact that allocatedLength is `roundUpToPowerOfTwo(length)` seems kinda insane but we can fix that in a different patch.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTable.h:
```diff
@@ -52,8 +51,18 @@ class Table : public ThreadSafeRefCounted<Table> {
     WTF_MAKE_NONCOPYABLE(Table);
     WTF_MAKE_TZONE_ALLOCATED(Table);
 public:
+    using MallocOps = VMMalloc;
```
Lets add `DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(WasmTable)` and use `WasmTableMalloc` instead.

kmiller68 reviewed (changes requested):
Your title seems wrong now that we're not removing MallocPtr anymore.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/498d230c306eb0fe45e24f9eb70d590728f9b34a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46668

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7e55d592b9db91b43318450bf6102b70879cd660. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46668

kmiller68 reviewed (approved):
r=me

kmiller68 commented:
I'm just gonna merge-queue since the only changes were comments.

kmiller68 commented:
Actually looks like you have some real test/build failures.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1cd5055032438f3fa6750f9068995bd5170df720. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46668

ast-hugger committed:
Require MallocPtr parameter type to be trivially destructible, correctly destroy FuncRefTable
https://bugs.webkit.org/show_bug.cgi?id=294022
rdar://152575154

Reviewed by Keith Miller.

This patch changes `MallocPtr` to require the parameter type `T` to be trivially destructible.

The usage of `MallocPtr` in `WasmTable` which was responsible for a callee leak is now
incompatible with that requirement. `WasmTable` is changed to use `std::unique_ptr`
with a custom deleter that correctly destroys the table contents.

Additionally, `VM` is changed to use a `FixedVector` instead of a `MallocPtr` to allocate
the exception fuzzing buffer.

* Source/JavaScriptCore/bytecode/ExpressionInfo.h:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::exceptionFuzzingBuffer):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::reallocate):
(JSC::Wasm::Table::grow):
(JSC::Wasm::ExternOrAnyRefTable::ExternOrAnyRefTable):
(JSC::Wasm::FuncRefTable::FuncRefTable):
(JSC::Wasm::FuncRefTable::~FuncRefTable): Deleted.
* Source/JavaScriptCore/wasm/WasmTable.h:
(JSC::Wasm::Table::StorageDeleter::operator()):
* Source/WTF/wtf/MallocPtr.h:

Canonical link: https://commits.webkit.org/297550@main
